plugins/experimental/client_allow_list/util.cc

```diff
@@ -0,0 +1,349 @@
+      for (std::size_t i = 1; i < list_of_matcher_idx_vectors.size(); ++i) {
+        *list_of_matcher_idx_vectors[i] = *list_of_matcher_idx_vectors[0];
+      }
+    }
+    if (sname_to_matcher_idxs.size() == 0) {
+      TSEmergency(PN ": YAML config file %s is empty", filespec.c_str());
+    }
+  } catch (const YAML::Exception &e) {
+    TSEmergency(PN ": YAML::Exception \"%s\" when parsing YAML config file %s", e.what(), filespec.c_str());
+  }
+}
+
+void
+Init::operator()(int n_args, char const *const *arg)
+{
+  if (n_args < 2) {
+    TSEmergency(PN ": must provide at least one plugin parameter");
+  }
+  if (2 == n_args) {
+    std::string_view first_arg(arg[1]);
+    if ((first_arg.size() > 5) && (first_arg.substr(first_arg.size() - 5) == ".yaml")) {
+      // Single parameter that is a yaml conifg file.
+      //
+      _yaml_process(arg[1]);
+      return;
+    }
+  }
+  // Otherwise the arguments are a list of name patterns that all client certs must match.
+  //
+  _process_name_args(n_args - 1, arg + 1);
+}
+
+void
+Init::_process_name_args(int n_names, char const *const *name)
+{
+  _Populator populator(this);
+
+  for (int i = 0; i < n_names; i++) {
+    if (!populator.add_cert_name(name[i])) {
+      TSEmergency(PN ": fatal error");
+    }
+  }
+  other_matcher_idxs = std::move(populator.matcher_idxs);
+  none_matcher_idxs  = other_matcher_idxs;
+}
+
+bool
+Init::_Populator::add_cert_name(std::string_view name)
+{
+  if (name.size() == 0) {
+    // Empty name, match nothing.
+    return true;
+  }
+  if (std::count(std::begin(name), std::end(name), '*') > 1) {
+    TSError(PN ": bad certificate name pattern %.*s", static_cast<int>(name.size()), name.data());
+    return false;
+  }
+  unsigned name_idx;
+  auto it = _init->_name_to_idx_map.find(name);
+  if (it != _init->_name_to_idx_map.end()) {
+    name_idx = it->second;
+    if (_idx_present_flag[name_idx]) {
+      TSError(PN ": duplicate name pattern %.*s", static_cast<int>(name.size()), name.data());
+      return false;
+    }
+  } else {
+    // New name pattern, so new matcher.
+
+    name_idx = matcher.size();
+    matcher.emplace_back();
+    matcher[name_idx]._CName = std::string(name.data(), name.size());
+
+    if (name.find("*") != std::string::npos) {
+      // Name pattern has wildcard, so translate it to a PCRE and compile it.
+
+      std::size_t si = 0, di = 1;
+      _init->_insert_tmp[0] = '^';
+
+      while (si < name.size()) {
+        if (_init->_insert_tmp.size() < (di + 8)) {
+          _init->_insert_tmp.resize(di + 128);
+        }
+        if ('.' == name[si]) {
+          std::memcpy(_init->_insert_tmp.data() + di, "\\.", 3);
+          di += 2;
+        } else if ('*' == name[si]) {
+          std::memcpy(_init->_insert_tmp.data() + di, ".{0,}", 5);
+          di += 5;
+        } else {
+          _init->_insert_tmp[di++] = name[si];
+        }
+        ++si;
+      }
+      _init->_insert_tmp[di++] = '$';
+      _init->_insert_tmp[di]   = '\0';
```

Comment:
This routine escapes only dots. If I understand the plugin correctly, though, we're validating CNs as well, which don't have to be DNS names. That means the CNs we're looking for might contain regex magic within them.

Also, it's worth noting that if an attacker can control the contents of these entries (you can imagine a self-service system where users can set the allowed CNs for certs connecting specifically with their domain, when ATS is hosting domains for multiple users), there's a fairly straightforward ReDoS attack. That might be fine under most threat models and working around it might be hard, so documenting the situation might be easier than fixing it if arbitrary regular expressions are allowed here. If the plan isn't to allow arbitrary regexen, this concern disappears when the input is sanitized.